<a href="https://colab.research.google.com/github/santhoshreddy39/Ineuron-ML-DL-Solutions/blob/master/Data_Cleaning_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN','londON_StockhOlm','Budapest_PaRis', 'Brussels_londOn'],
                   'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
                   'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )','12. Air France', '"Swiss Air"']})

In [3]:
print(df)

            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"


In [4]:
# Replacing NaN values with previous value + 10.
df['FlightNumber'].fillna(value=df['FlightNumber'][0]+10,limit=1,inplace=True)  
df['FlightNumber'].fillna(value=df['FlightNumber'][2]+10,limit=1,inplace=True)
print(df)

            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN       10055.0            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis       10075.0          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"


1. Some values in the the FlightNumber column are missing. These numbers are meant to increase by 10 with each row so 10055 and 10075 need to be put in place. Fill in these missing numbers and make the column an integer column (instead of a float column).

In [5]:
# Converting FlightNumber column from float to interger values.
df['FlightNumber']=df['FlightNumber'].astype(np.int64)
df['FlightNumber'].dtype

dtype('int64')

2. The From_To column would be better as two separate columns! Split each string on the underscore delimiter _ to give a new temporary DataFrame with the correct values. Assign the correct column names to this temporary DataFrame.

In [6]:
temp_df=pd.DataFrame(df['From_To'].str.split("_",1).to_list(),columns=['From','To'])
print(temp_df)

       From         To
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn


3. Notice how the capitalisation of the city names is all mixed up in this temporary DataFrame. Standardise the strings so that only the first letter is uppercase (e.g. "londON" should become "London".)

In [7]:
temp_df['From']=temp_df['From'].str.capitalize()
temp_df['To']=temp_df['To'].str.capitalize()
print(temp_df)

       From         To
0    London      Paris
1    Madrid      Milan
2    London  Stockholm
3  Budapest      Paris
4  Brussels     London


4. Delete the From_To column from df and attach the temporary DataFrame from the previous questions.

In [8]:
print(df)

            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"


In [9]:
df.drop(['From_To'],inplace=True,axis=1)
print(df)

   FlightNumber  RecentDelays              Airline
0         10045      [23, 47]               KLM(!)
1         10055            []    <Air France> (12)
2         10065  [24, 43, 87]  (British Airways. )
3         10075          [13]       12. Air France
4         10085      [67, 32]          "Swiss Air"


In [10]:
df=pd.concat([temp_df,df],axis=1,sort=False)
print(df)

       From         To  FlightNumber  RecentDelays              Airline
0    London      Paris         10045      [23, 47]               KLM(!)
1    Madrid      Milan         10055            []    <Air France> (12)
2    London  Stockholm         10065  [24, 43, 87]  (British Airways. )
3  Budapest      Paris         10075          [13]       12. Air France
4  Brussels     London         10085      [67, 32]          "Swiss Air"


5. In the RecentDelays column, the values have been entered into the DataFrame as a list. We would like each first value in its own column, each second value in its own column, and so on. If there isn't an Nth value, the value should be NaN.

In [11]:
delay=pd.DataFrame(df['RecentDelays'].to_list(),columns=['delay_1','delay_2','delay_3'])
delay

delay_1  delay_2  delay_3
0     23.0     47.0      NaN
1      NaN      NaN      NaN
2     24.0     43.0     87.0
3     13.0      NaN      NaN
4     67.0     32.0      NaN

In [12]:
df.drop(['RecentDelays'],inplace=True,axis=1)
print(df)

       From         To  FlightNumber              Airline
0    London      Paris         10045               KLM(!)
1    Madrid      Milan         10055    <Air France> (12)
2    London  Stockholm         10065  (British Airways. )
3  Budapest      Paris         10075       12. Air France
4  Brussels     London         10085          "Swiss Air"


In [13]:
df.insert(loc=3,column='delay_1',value=delay['delay_1'])
df.insert(loc=4,column='delay_2',value=delay['delay_2'])
df.insert(loc=5,column='delay_3',value=delay['delay_3'])
print(df)

       From         To  FlightNumber  ...  delay_2  delay_3              Airline
0    London      Paris         10045  ...     47.0      NaN               KLM(!)
1    Madrid      Milan         10055  ...      NaN      NaN    <Air France> (12)
2    London  Stockholm         10065  ...     43.0     87.0  (British Airways. )
3  Budapest      Paris         10075  ...      NaN      NaN       12. Air France
4  Brussels     London         10085  ...     32.0      NaN          "Swiss Air"

[5 rows x 7 columns]
